### ISSUE TITLE
Disallow implicit typecast of void values to boolean

### ISSUE THREAD

**Issue by johnsparrowuk** [palantir/tslint#719]
I accidentally typed `if (alert("something")) ...` the other day, and was surprised neither TS itself or TSLint picked this up.

Is it possible to have a rule for this? The language spec (section 3.2.5) says void is a super-type of null and undefined. So it is distinct.

**Comment by adidahiya** [comment]
What exactly do you mean by "evaluation" here? If you want to forbid this in conditional expressions, I think #253 is what you want

In general, rules based on type introspection are blocked on #680

**Comment by johnsparrowuk** [comment]
#253 is closer I think, but I mean something more specific. Just disallowing logical expressions involving 'void' types.

**Comment by adidahiya** [comment]
yeah, #253 actually covers that (the title says "logical expressions")

**Comment by adidahiya** [comment]
oh wait, it sounds like you want to disallow:

```ts
if (alert("something") ...
```

but still allow this?

```ts
if ("something) ...
```

**Comment by johnsparrowuk** [comment]
Yes, as I understand it #253 would disallow anything that isn't `boolean`. That's useful but very strict, I'm suggesting just disallowing `void` types. I can't see a use-case for evaluating a void value, since it will always be null.

**Comment by adidahiya** [comment]
Cool, sounds good, I'll re-categorize this as a feature request. Are you ok with renaming this to "disallow implicit typecast of void values to boolean"?

**Comment by johnsparrowuk** [comment]
Great, thank you.

**Comment by andy-hanson** [comment]
Would #2033 help? That would also cover `if (foo())` where `foo()` returns, say, `{ x: number }`, and isn't possibly `undefined`.

**Comment by andy-hanson** [comment]
@adidahiya I think this can be closed now.
